### PR TITLE
Class distribution for classification tasks 

### PR DESCRIPTION
**Pull request by paxcema**
## Why
For classification tasks, we need Lightwood to return the complete class distribution probabilities for certain scenarios downstream, like using XAI techniques like LIME, or running MindsDB in the OpenML AutoML benchmark suite.

## How
Added support for returning class probabilities in all mixers (SkLearn and Boost by calling `.predict_proba(X)`, NnMixer by adding non-calibrated softmax to the one-hot encoder). Predictions are now sent to native as a dictionary.

### CODE REVIEW

File: lightwood/encoders/categorical/onehot.py [hunk 1]
```diff
@@ -60,9 +63,22 @@ def encode(self, column_data):
...
+                del(vector[UNCOMMON_TOKEN])
+                probs.append(softmax(vector).tolist())
+
+        if self.predict_proba:
+            # UNK not included in class_map nor belief distribution
+            if UNCOMMON_TOKEN != 0:
+                raise Exception("Uncommon token should be the first assigned token in the vocabulary, aborting.")
+            class_map = deepcopy(self._lang.index2word)
+            del(class_map[UNCOMMON_TOKEN])
+            return ret, probs, {k-1: v for k, v in class_map.items()}
```
Comment (George3d6) [inline lightwood/encoders/categorical/onehot.py:82]: `{k-1: v for k, v in class_map.items() if k != UNCOMMON_TOKEN}`

should solve the need for a potentially expensive deepcopy (since done very often) and make the code less confuing (since that `del` is no longer needed)

File: lightwood/encoders/categorical/onehot.py [hunk 2]
```diff
@@ -1,6 +1,8 @@
 import torch
-from lightwood.encoders.text.helpers.rnn_helpers import Lang
 import numpy as np
+from copy import deepcopy
+from scipy.special import softmax
+from lightwood.encoders.text.helpers.rnn_helpers import Lang
```
Comment (George3d6) [inline lightwood/encoders/categorical/onehot.py:5]: Should we really reuse this here ? Isn't it better to create a helper that is inherited by `Lang` and get rid of things like the `UNCOMMON_TOKEN` deletion (which seems to be uncecessary for the purposes of this encoder) ?

File: lightwood/mixers/boost.py [hunk 3]
```diff
@@ -91,6 +91,11 @@ def _predict(self, when_data_source, include_extra_data=False):
                 else:
                     predictions[target_col_name]['predictions'] = list(self.targets[target_col_name]['model'].predict(X))
 
+                    # add distribution belief if the flag was set in the target encoder
+                    if hasattr(self.encoders[target_col_name], 'predict_proba') and self.encoders[target_col_name].predict_proba:
```
Comment (George3d6) [inline lightwood/mixers/boost.py:95]: `if getattr(self.encoders[target_col_name], 'predict_proba', False) `

File: lightwood/mixers/sklearn_mixer.py [hunk 4]
```diff
@@ -183,6 +183,12 @@ def _predict(self, when_data_source, include_extra_data=False):
                         self.targets[target_col_name]['model'].predict(X)
                     )
 
+
+                    # add distribution belief if the flag was set in the target encoder
+                    if hasattr(self.encoders[target_col_name], 'predict_proba') and self.encoders[target_col_name].predict_proba:
```
Comment (George3d6) [inline lightwood/mixers/sklearn_mixer.py:188]: if getattr(self.encoders[target_col_name], 'predict_proba', False)

File: lightwood/api/data_source.py [hunk 5]
```diff
@@ -457,7 +457,16 @@ def get_decoded_column_data(self, column_name, encoded_data, decoder_instance=No
                     You should not decode before having encoding at least once
                     """)
             decoder_instance = self.encoders[column_name]
-        decoded_data = decoder_instance.decode(encoded_data)
+
+        decoded_data = {}
+        if hasattr(decoder_instance, 'predict_proba') and decoder_instance.predict_proba:
```
Comment (George3d6) [inline lightwood/api/data_source.py:462]: if getattr(self.encoders[target_col_name], 'predict_proba', False)

File: lightwood/mixers/nn.py [hunk 6]
```diff
@@ -447,6 +447,10 @@ def _predict(self, when_data_source, include_extra_data=False):
             if include_extra_data:
                 predictions[output_column]['encoded_predictions'] = output_trasnformed_vectors[output_column]
 
+            if 'predict_proba' in decoded_predictions:
+                predictions[output_column]['class_distribution'] = decoded_predictions['predict_proba']
+                predictions[output_column]['class_labels'] = decoded_predictions['class_labels']
```
Comment (George3d6) [inline lightwood/mixers/nn.py:452]: Why not keep consistent naming ? i.e call `class_distribution` here `predict_proba` or vice versa, replace `predict_proba` key with `class_distribution`

**Review by George3d6 — CHANGES_REQUESTED** [review]

File: lightwood/encoders/categorical/onehot.py (hunk 2 above)
Comment (paxcema) [reply lightwood/encoders/categorical/onehot.py:5]: Notice that `Lang` usage isn't introduced here. I agree we could probably do this, however I'm not sure if it should be part of this PR as it may impact accuracy in the categorical AE.

Comment (George3d6) [reply lightwood/encoders/categorical/onehot.py:4]: Ah, ok, my bad, will make an issue about this.
